CURATOR-599 Configurable ZookeeperFactory by ZKClientConfig

liran2000 opened the pull request:
Option to use ZooKeeper client config.

This seems mandatory for using zookeeper.request.timeout for preventing
the potential race condition of hanging indefinitely, as described at
the ticket.

liran2000 commented:
it depends on ZooKeeper 3.6.1+ which recently updated.

It is passing locally, but failing in the CI build in GitHub, can anyone please help with that ?

@eolivelli @cammckenzie @tisonkun 

Possibly it is since Maven running there without 'clean' step, but I am not sure.
(mvn install -DskipTests=true -Dmaven.javadoc.skip=true -B -V)
Failing on java.lang.NoSuchMethodError: org.apache.zookeeper.admin.ZooKeeperAdmin: method <init>(Ljava/lang/String;ILorg/apache/zookeeper/Watcher;ZLorg/apache/zookeeper/client/ZKClientConfig;)V not found

eolivelli commented:
taking a look now

eolivelli commented:
This is an error in the compatibility tests for 3.5.
Basically those tests are there in order to prove that even if you have only ZK 3.5 in the classpath the basic features of Curator work

![image](https://user-images.githubusercontent.com/9469110/124885499-9ec90600-dfd3-11eb-95e5-b012e5c104e0.png)

eolivelli commented:
Probably we have to detect the availability of that method and fallback to legacy code in case we are on an older version of ZK.

eolivelli commented:
@cammckenzie @Randgalt do you mind taking a look as well?

eolivelli reviewed (approved):
LGTM

cammckenzie reviewed (changes requested):
Some minor comments, otherwise, LGTM

tisonkun reviewed (approved):
I second @cammckenzie 's suggestion and other parts LGTM.

Randgalt commented on curator-framework/src/main/java/org/apache/curator/framework/imps/CuratorFrameworkImpl.java:
```diff
@@ -200,23 +232,39 @@ public QuorumVerifier getCurrentConfig()
-    private ZookeeperFactory makeZookeeperFactory(final ZookeeperFactory actualZookeeperFactory)
+    private ZookeeperFactory makeZookeeperFactory(final ZookeeperFactory actualZookeeperFactory, ZKClientConfig zkClientConfig)
     {
+    	if (actualZookeeperFactory instanceof ConfigurableZookeeperFactory) {
```
I'm -1 on using `instanceof`. Wouldn't it be better to add a `default` method to `ZookeeperFactory` that takes the `ZKClientConfig`? It could be optional and forward to the older versions when not present. The default version in `DefaultZookeeperFactory` would throw an exception.

liran2000 replied:
Understood. Done.

liran2000 commented:
Thank you all for the reviews and info, I addressed the comments. Can you please have another review and merge it ?

@eolivelli @cammckenzie @tisonkun @Randgalt

liran2000 commented on curator-framework/src/test/java/org/apache/curator/framework/imps/TestFramework.java:
```diff
@@ -1070,10 +1077,28 @@ public void testCreateModes() throws Exception
 
             stat = client.checkExists().forPath(realPath);
             assertNull(stat);
+            
```
Done

eolivelli commented on curator-framework/src/test/java/org/apache/curator/framework/imps/TestFramework.java:
```diff
@@ -50,26 +59,24 @@
 @Tag(CuratorTestBase.zk35TestCompatibilityGroup)
 public class TestFramework extends BaseClassForTests
 {
+	private final Logger log = LoggerFactory.getLogger(getClass());
```
Usually loggers are static fields
`private static final Logger log = LoggerFactory.getLogger(TestFramework.class);`

liran2000 replied:
I wrote it exactly as already implemented at the same package, to be consistent:

https://github.com/apache/curator/blob/b9dc9bd4056dba18c29a4f3127df43b930449abe/curator-framework/src/test/java/org/apache/curator/framework/imps/TestFrameworkBackground.java#L56

https://github.com/apache/curator/blob/b9dc9bd4056dba18c29a4f3127df43b930449abe/curator-framework/src/test/java/org/apache/curator/framework/imps/TestFrameworkEdges.java#L78

eolivelli replied:
sorry.
you are right.
I forgot that we are in Curator and we have different conventions.

thanks for checking out

eolivelli reviewed (approved):
LGTM

waiting for @cammckenzie final feedback

cammckenzie reviewed (approved):
There a couple of minor formatting issues, but we can fix these when we merge.

LGTM, thanks @liran2000

liran2000 commented:
@eolivelli thank you, can you please merge it and then resolve the Jira issue ?

eolivelli commented on curator-framework/src/test/java/org/apache/curator/framework/imps/TestFramework.java:
```diff
@@ -1073,7 +1080,41 @@ public void testCreateModes() throws Exception
+            assertEquals(zookeeperRequestTimeout, client.getZookeeperClient().getZooKeeper().getClientConfig().getProperty(ZKClientConfig.ZOOKEEPER_REQUEST_TIMEOUT));
+            
+        } catch (NoSuchMethodError e) {
+			log.debug("NoSuchMethodError: ", e);
```
Can you please fix formatting here?

Otherwise I will do it as follow up.
I will merge this patch today.

We are also going to cut a release as soon as possible

liran2000 replied:
Sure, done.
BTW, at my IDE this formatting, formatting was fine, depends on IDE config, I will notice it.

liran2000 replied:
@eolivelli thanks, can be merged and then resolve the Jira issue

eolivelli commented:
probably you are linking the wrong JIRA
https://issues.apache.org/jira/browse/CURATOR-599 "Hanging indefinitely on some scenarios since zookeeper.request.timeout cannot be configured"

eolivelli commented:
I will be happy to commit once we are pointing to the right JIRA

liran2000 commented:
@eolivelli  it is the right JIRA issue. The issue should be resolved when zookeeper.request.timeout can be configured - with this commit, it can be configured, then the issue should not occur. It was Hanging indefinitely, and when ZooKeeper request timeout is configured - it will get a timeout as expected, and not hang indefinitely . See below line that added at the test to configure it, and more details are there in the JIRA itself. If something is not clear you can let me know.

https://github.com/apache/curator/pull/391/files#diff-b423c4cc97cc1282306109606a1ab1060ce818cd1b04d6ea7c7493ca657bbf5dR1095

zkClientConfig.setProperty(ZKClientConfig.ZOOKEEPER_REQUEST_TIMEOUT, zookeeperRequestTimeout);

https://issues.apache.org/jira/browse/CURATOR-599 "Hanging indefinitely on some scenarios since zookeeper.request.timeout cannot be configured"

liran2000 commented:
@eolivelli any other clarification needed for the comment above ?
when zookeeper.request.timeout can be configured and used, it can resolve the issue at Jira:
CURATOR-599 - Hanging indefinitely on some scenarios since zookeeper.request.timeout cannot be configured

eolivelli commented:
committed to master branch

liran2000 commented:
Thank you!
